src/webapi/events/drag.rs

```diff
@@ -0,0 +1,697 @@
+#[reference(subclass_of(Event, MouseEvent))]
+pub struct DragOverEvent(Reference);
+
+impl IEvent for DragOverEvent {}
+
+impl IUiEvent for DragOverEvent {}
+
+impl IMouseEvent for DragOverEvent {}
+
+impl IDragEvent for DragOverEvent {}
+
+impl ConcreteEvent for DragOverEvent {
+    const EVENT_TYPE: &'static str = "dragover";
+}
+
+/// The dragstart event is fired when the user starts dragging an element or text selection.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/Events/drop)
+// https://www.w3.org/TR/html51/editing.html#eventdef-global-drop
+#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
+#[reference(instance_of = "DragEvent")]
+#[reference(subclass_of(Event, MouseEvent))]
+pub struct DropEvent(Reference);
+
+impl IEvent for DropEvent {}
+
+impl IUiEvent for DropEvent {}
+
+impl IMouseEvent for DropEvent {}
+
+impl IDragEvent for DropEvent {}
+
+impl ConcreteEvent for DropEvent {
+    const EVENT_TYPE: &'static str = "drop";
+}
+
+/// The dragend event is fired when a drag operation is being ended (by releasing a mouse button or
+/// hitting the escape key).
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/Events/dragend)
+// https://www.w3.org/TR/html51/editing.html#eventdef-global-dragend
+#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
+#[reference(instance_of = "DragEvent")]
+#[reference(subclass_of(Event, MouseEvent))]
+pub struct DragEndEvent(Reference);
+
+impl IEvent for DragEndEvent {}
+
+impl IUiEvent for DragEndEvent {}
+
+impl IMouseEvent for DragEndEvent {}
+
+impl IDragEvent for DragEndEvent {}
+
+impl ConcreteEvent for DragEndEvent {
+    const EVENT_TYPE: &'static str = "dragend";
+}
+
+/// The DataTransfer object is used to hold the data that is being dragged during a drag and drop
+/// operation.
+/// It may hold one or more data items, each of one or more data types.
+/// For more information about drag and drop, see HTML Drag and Drop API.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer)
+// https://www.w3.org/TR/html51/editing.html#the-datatransfer-interface
+#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
+#[reference(instance_of = "DataTransfer")]
+pub struct DataTransfer( Reference );
+impl DataTransfer {
+    /// Gets the type of drag-and-drop operation currently selected or sets the operation to a new
+    /// type. The value must be none, copy, link or move.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/dropEffect)
+    pub fn drop_effect( &self ) -> DropEffect {
+        let v: String = js!(
+            return @{self.as_ref()}.dropEffect;
+        ).try_into().unwrap();
+        match v.as_ref() {
+            "copy" => DropEffect::Copy,
+            "move" => DropEffect::Move,
+            "link" => DropEffect::Link,
+            "none" => DropEffect::None,
+            other => panic!("Expected valid dropEffect value, got {:?}", other),
```

Comment:
If it's true that this can return non-standard values then something like this would be good:

```
pub enum DropEffect {
    None,
    Copy,
    Move,
    Link,
    Other( OtherDropEffect )
}
impl DropEffect {
    fn as_str( &self ) -> &str { ... }
}
struct OtherDropEffect( String );
```

Basically prevent direct access to the string. But I'd really like if someone could investigate if this is really necessary.

Reply:
It's more future proof in case we have to add more variants.

(I also forgot to mention that if we add the the `Other` variant then it should be `#[doc(hidden)]` and the whole enum should be explicitly marked as non-exhaustive in the docs.)

Let's assume we give the user direct access to the string, so they write the following code:

```
match drop_effect {
    DropEffect::Copy => { ... },
    DropEffect::Other( other ) if other == "foobar" => { ... },
    _ => { ... }
}
```

Now we decide to add `DropEffect::Foobar`. The user updates, and their code is broken. If we don't give the user a direct access to that variant then they have to write their code like this:

```
match drop_effect {
    DropEffect::Copy => { ... },
    other => if other.as_str() == "foobar" { ... },
    _ => { ... }
}
```

And now their code still works after we add `DropEffect::Foobar`.

Reply:
Love it, thanks I'll implement this

Reply:
@koute here are a few points regarding this, to make sure we are on the same page:
- we only needed an extra variant for the `DataTransferItemKind` enum only, not for the other two `DropEffect` and `EffectAllowed` (tested + googled around)
- seems like we'd want to use [#[non_exhaustive]](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md), but it is an experimental Rust feature at the moment. I added a TODO to use it in the future.
- I made the type `OtherKind` public, since variant cannot leak private types ([E0446](https://doc.rust-lang.org/error-index.html#E0446))
- I made the string field private so that the user must use `.as_str()`
- `.as_str()` has to be used on the variable of type `DataTransferItemKind`, I'm assuming that's what you meant in your example

I used it like so in the example project:
```
let kind_str = match x.kind() {
    DataTransferItemKind::String => String::from("a string"),
    DataTransferItemKind::File => String::from("a file"),
    DataTransferItemKind::Other(_) if x.kind().as_str() == "other" => String::from("an expected other kind"),
    DataTransferItemKind::Other(_) => format!("an unexpected other kind \"{}\"", x.kind().as_str()),
};
```

Reply:
Sounds good to me! However, we shouldn't use the `DataTransferItemKind::Other` explicitly in the example - it's marked `#[doc(hidden)]` after all! The whole point of it is that the user should never ever access that variant directly in any way.

So using your example:

```
let kind_str = match x.kind() {
    DataTransferItemKind::String => String::from("a string"),
    DataTransferItemKind::File => String::from("a file"),
    item_kind if item_kind.kind().as_str() == "other" => String::from("an expected other kind"),
    item_kind => format!("an unexpected other kind \"{}\"", item_kind.kind().as_str()),
};
```

Actually, one more thing - let's rename the `Other` variant to, I don't know, `__Other` or something like that, to make it less likely that someone will use it by mistake.